Sys management agent

akramtexas opened the pull request:
First of all, this PR is **tagged and associated with** [Issue #497](https://github.com/edgexfoundry/edgex-go/issues/497)

`This PR essentially has two major parts:`

1. Bringing the new System Management Agent functionality into master
2. The refactoring done to split business logic from routing / INFRA concerns. 

- Details regarding the former can be [found here](https://wiki.edgexfoundry.org/display/FA/Phase+1%2C+System+Management+Delhi+Release)
- And details regarding the latter can be [found here](https://github.com/edgexfoundry/edgex-go/issues/457).

drasko commented on cmd/sys-mgmt/commands.go:
```diff
@@ -0,0 +1,16 @@
+package main
+
+// Start() begins execution of the agent in the provided context, if not already
```
Dead code - please remove

drasko commented on cmd/sys-mgmt/commands.go:
```diff
@@ -0,0 +1,16 @@
+package main
```
Missing copyright header.

drasko commented on cmd/sys-mgmt/config/agent.go:
```diff
@@ -0,0 +1,14 @@
+package config
```
Copyright header.

drasko commented on cmd/sys-mgmt/config/agent.go:
```diff
@@ -0,0 +1,14 @@
+	started   chan struct{}
+	startOnce sync.Once     // start only once
+	stopped   chan struct{} // requests shutdown
+	stopOnce  sync.Once     // start only once
```
Fix comment - stop.

drasko commented on cmd/sys-mgmt/config/const.go:
```diff
@@ -0,0 +1,69 @@
+	ZeroMQAddressPort          string
+}
+
+var configuration ConfigurationStruct = ConfigurationStruct{} //  Needs to be initialized before used
```
Should be initialized in `main()` probably, in `cmd` somewhere.

tsconn23 replied:
Why does this file exist at all? No other service has a cmd/config directory with go code in it??

drasko commented on cmd/sys-mgmt/config/const.go:
```diff
@@ -0,0 +1,69 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+package config
```
Why is this file called `const.go`, when all we have are variables?

drasko commented on cmd/sys-mgmt/main.go:
```diff
@@ -0,0 +1,105 @@
+	startHttpServer(errs, agent.Configuration.ServicePort)
+
+	// Time it took to start service
+	fmt.Printf("Service started in: "+time.Since(start).String(), "")
```
No `printf` on stdout - use logger.

tsconn23 replied:
@akramtexas To @drasko's point, follow the patterns available in other services where main uses LoggingClient to log. In cases where you have to log before bootstrap is complete (or it has failed) see `func logBeforeInit(err error)`

drasko commented on cmd/sys-mgmt/toml_test.go:
```diff
@@ -0,0 +1,22 @@
+//
+// Copyright (c) 2018
+// Cavium
+//
```
Add Dell

drasko commented on cmd/sys-mgmt/services.go:
```diff
@@ -0,0 +1,10 @@
+package main
```
Copyright header missing.

drasko commented on docker/Dockerfile.sys-management-agent:
```diff
@@ -0,0 +1,30 @@
+#
+# Copyright (c) 2018
+# Cavium
+#
```
Add Dell

drasko commented on internal/system/agent/const.go:
```diff
@@ -0,0 +1,69 @@
+	ZeroMQAddressPort          string
+}
+
+var configuration ConfigurationStruct = ConfigurationStruct{} //  Needs to be initialized before used
```
Not a right place for initializations. Should be in `main()` somewhere.

tsconn23 replied:
@akramtexas There's a bug here. The `var configuration` line should be removed and any references to it pointed instead to the Configuration variable globally available in the agent package. See line 35:init.go

drasko commented on internal/system/agent/init.go:
```diff
@@ -0,0 +1,195 @@
+				}
+			} else {
+				// Setup Logging
+				fmt.Printf("LOGGING >> TODO: Enable Remote Logging")
```
No `Printf()`s - use logger.

drasko commented on internal/system/agent/services.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+const (
+	START   = "start"
```
Using capitals for const is not Go idiomatic.

drasko replied:
Additionally - do you need string consts? This is handled more elegantly with [iota](https://splice.com/blog/iota-elegant-constants-golang/)

drasko commented on internal/system/agent/services.go:
```diff
@@ -0,0 +1,44 @@
+	// Loop through requested action, along with respectively-supplied parameters.
+	for _, service := range services {
+		// TODO: Logging is placeholder for actual functionality to be implemented.
+		fmt.Printf("OPERATION >> About to {%v} the service {%v} with parameters {%v}.\n", action, service, params)
```
No printfs on stdout - please log

drasko commented on internal/system/agent/services.go:
```diff
@@ -0,0 +1,44 @@
+		for _, metric := range metrics {
+			// TODO: Logging is placeholder for actual functionality to be implemented.
+			// TODO: What constitutes metrics. How do services provide those metrics?
+			fmt.Printf("METRIC >> Fetching the metric {%v} for the service {%v}...\n", metric, service)
```
Use logger

drasko commented on internal/system/agent/services_test.go:
```diff
@@ -0,0 +1,72 @@
+package agent
```
Missing copyright header.

drasko commented on internal/system/agent/services.go:
```diff
@@ -0,0 +1,44 @@
+package agent
```
Missing copyright header.

drasko commented:
@akramtexas first review run finished. Overall very good, great work.

tsconn23 commented on internal/constants.go:
```diff
@@ -20,3 +20,4 @@ const ExportClientServiceKey = "edgex-export-client"
 const ExportDistroServiceKey = "edgex-export-distro"
 const SupportLoggingServiceKey = "edgex-support-logging"
 const SupportNotificationsServiceKey = "edgex-support-notifications"
+const SystemManagementAgentServiceKey = "sys-mgmt"
```
I suggest "edgex-sys-mgmt-agent"

tsconn23 commented:
@akramtexas See if you can resolve the build failure and then I'll take a look.

akramtexas commented:
@drasko Thank you very much for making the time to review, for providing feedback, as well as for the kind words 👍  I'll shortly follow up on the provided feedback.

akramtexas commented:
@tsconn23 Found the issue, and looking into fixing it now:

`15:58:02 [91mmake: *** No rule to make target 'cmd/docker-sys-management-agent/docker-sys-management-agent'. `

jpwhitemn commented:
Nice work - thanks @tsconn23 .  Good first steps in system management @akramtexas .

akramtexas commented:
Thanks @jpwhitemn - Yay, the _edgex-go-master-verify_ build has finished successfully 💯 Now on to the ARM build, which (I'm reckoning) will have a (far) longer wait in the Jenkins queue (pipeline).

tsconn23 commented on Makefile:
```diff
@@ -11,10 +11,10 @@
 .PHONY: $(DOCKERS)
 
-MICROSERVICES=cmd/export-client/export-client cmd/export-distro/export-distro cmd/core-metadata/core-metadata cmd/core-data/core-data cmd/core-command/core-command cmd/support-logging/support-logging cmd/support-notifications/support-notifications
+MICROSERVICES=cmd/export-client/export-client cmd/export-distro/export-distro cmd/core-metadata/core-metadata cmd/core-data/core-data cmd/core-command/core-command cmd/support-logging/support-logging cmd/support-notifications/support-notifications cmd/sys-mgmt/sys-mgmt
```
Should `cmd/sys-mgmt` instead be `cmd/sys-mgmt-agent`?

tsconn23 commented on Makefile:
```diff
@@ -47,6 +47,9 @@ cmd/support-logging/support-logging:
 	$(GO) build $(GOFLAGS) -o $@ ./cmd/support-notifications
 
+cmd/sys-mgmt/sys-mgmt:
+	$(GO) build $(GOFLAGS) -o $@ ./cmd/sys-mgmt
```
Should `cmd/sys-mgmt` instead be `cmd/sys-mgmt-agent`?

tsconn23 commented on Makefile:
```diff
@@ -120,3 +123,11 @@ docker_support_notifications:
 		-t edgexfoundry/docker-support-notifications-go:$(GIT_SHA) \
 		-t edgexfoundry/docker-support-notifications-go:$(VERSION)-dev \
 		.
+
```
Same `sys-mgmt-agent`

tsconn23 commented on cmd/sys-mgmt/res/configuration-docker.toml:
```diff
@@ -0,0 +1,50 @@
+ConsulProfilesActive = 'docker;go'
```
Should not be 'docker;go' because of the bug you helped fix.

tsconn23 commented on cmd/sys-mgmt/res/configuration-docker.toml:
```diff
@@ -0,0 +1,50 @@
+MsgPubType = 'zero'
+ServicePort = 48080
+ServiceTimeout = 5000
+ServiceAddress = 'edgex-core-data'
```
`edgex-core-data`???

tsconn23 commented on cmd/sys-mgmt/res/configuration-docker.toml:
```diff
@@ -0,0 +1,50 @@
+AppOpenMsg = 'This is the Core Data Micro Service'
+FormatSpecifier = '%(\\d+\\$)?([-#+ 0(\\<]*)?(\\d+)?(\\.\\d+)?([tT])?([a-zA-Z%])'
+MsgPubType = 'zero'
+ServicePort = 48080
```
This is certainly the wrong service port

tsconn23 commented on cmd/sys-mgmt/res/configuration-docker.toml:
```diff
@@ -0,0 +1,50 @@
+MongoDBPort = 27017
+MongoDBConnectTimeout = 5000
+ConsulHost = 'edgex-core-consul'
+ConsulCheckAddress = 'http://edgex-core-data:48080/api/v1/ping'
```
This is all just cut-and-pasted as is the non-docker configuration file. Please review these with an eye toward what your service actually needs. If it doesn't need anything right now, then remove all this stuff.

tsconn23 commented on cmd/sys-mgmt/res/configuration.toml:
```diff
@@ -0,0 +1,50 @@
+ConsulProfilesActive = 'go'
```
See above comments RE: cut-and-paste

tsconn23 commented on internal/system/agent/const.go:
```diff
@@ -0,0 +1,69 @@
+package agent
+
+type ConfigurationStruct struct {
+	ConsulProfilesActive       string
```
Does your service really need all of this for Configuration? Only include what you need. You can add back later.

akramtexas commented:
@tsconn23 This is excellent feedback, thanks! I had debated pulling over only specific things for this new service (but thereby possibly risking missing some scaffolding and / or "context" that might later get missed or overlooked later on). But your comment has made me rethink that strategy and I will immediately remove all scaffolding and "context" cruft and make this as lean as possible 👍

akramtexas commented:
@tsconn23 and @drasko Yep, making a note to incorporate your collective comment about using the logging capability. Where I was initially coming from was the consideration that the SMA may not be able to depend on the logging service being available for it to use (remotely). Hence my removing all references to LoggingClient. But given the feedback, I realize that this particular concern of mine is not warranted for the SMA operational environment (locally). Good point, thanks.

akramtexas commented:
- @tsconn23 The fourth check (i.e. the ARM build), too, has passed. 
- Rebasing has made this branch "current" and merge-able with master branch. 
- Per your recommendation, let's please squash all commits related to this single issue into a single commit, after the PR gets approved, but before we merge, thanks.

akramtexas commented:
@tsconn23 Per your recommendation, tagging you with the update that the sys-mgmt-agent service  is now running without issue (I tested it locally). Also, stripped out the database logic until its need is warranted.

akramtexas commented:
- Previously, I had intended to mention (in my prior comment above) the observation that the SMA service was taking approximately 31 seconds to start up, which struck me as odd (while doing due diligence during local testing), but inadvertently missed making mention of it, which is something I regret. At any rate, I've since then added logic to exit the loop if the dependencies have been satisfied at the launch of the SMA service. 
 
- Also added a ping endpoint, along with a pingHandler (Great suggestion, Trevor). FWIW, I also added a new ServicePort for the SMA, tentatively assigning it the number 48090.
 
- The SMA service launches as part of `make run`, with console logging like so:

```
INFO: 2018/08/28 13:52:02 Starting edgex-sys-mgmt-agent 0.7.0 
INFO: 2018/08/28 13:52:02 This is the System Management Agent Service
INFO: 2018/08/28 13:52:02 Service started in: 986.861µs
INFO: 2018/08/28 13:52:02 Listening on port: 48090
```

- The newly-added ping endpoint (http://localhost:48090/api/v1/ping) returns the expected "pong" response.

- Again, I've tentatively assigned port number `48090` as the ServicePort for the SMA. This can be changed if there is a need to do so, although I don't see any conflicts... (The currently allocated ports I see are: `48081, 48080, 48082, 48061, 48060, 48071, 48070, 48075, 49990`)

@tsconn23 I will tag you once all four checks have passed (three have, thus far). Thanks for your patience, as well as for great feedback 👍

akramtexas commented:
- While working on a different branch `micro_service_mgmt_api`, I came across three straggling `fmt.Printf` style statements that will also be replaced by suitable `LoggingClient.Info` style statements, for example this one will be replaced:

`fmt.Printf("CONFIG >> Fetching the configuration for the service {%v}!\n", service)`

akramtexas commented:
@tsconn23 Sounds good, and thanks again for the excellent feedback. Yep, I was planning on getting with @JPWKU and squash the commits down to one (prior to merging, as discussed earlier). I sure will tag you as and when we require another approval (Appreciated the heads up in this regard.) 👍

akramtexas commented:
@tsconn23 Currently working with @JPWKU who is helping me resolve an issue with merge conflicts during rebasing, specifically with `git rebase -i HEAD~50`

akramtexas commented:
@tsconn23 As promised, I'll tag you once all four triggered checks (underway) have passed after pushing the rebased (squashed commits). Completed the merging conflicts encountered during rebasing, specifically with `git rebase -i HEAD~49` - Thanking @JPWKU for stellar help with Git wizardry in resolving merge issues in that process 👍

akramtexas commented:
@tsconn23 All triggered checks have passed after I had earlier on pushed the rebased (squashed commits) to my branch for this PR. At this time, I require another approval: Tagging you as promised, thanks 👍
